File: medium/validparenthesesstring.cpp
#include <iostream>

class Solution {
public:
    bool checkValidString(string s) {
        int min = 0, max = 0;

        for(char c : s) {
            if(c == '(') {
                min += 1;
                max += 1;
            } else if(c == ')') {
                min -= 1;
                max -= 1;
            } else {
                min -= 1;
                max += 1;
            }
            if(max < 0) return false;
            if(min < 0) min = 0;
        }

        return min == 0;
    }
};

File: python/medium/validparenthesesstring.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        mini = 0
        maxi = 0

        for c in s:
            if c == '(':
                mini += 1
                maxi += 1
            elif c == ')':
                mini -= 1
                maxi -= 1
            else:
                mini -= 1
                maxi += 1

            if maxi < 0:
                return False
            if mini < 0:
                mini = 0
            
        return mini == 0